core/src/toga/widgets/table.py

```diff
@@ -1,252 +1,327 @@
-        self._accessors = build_accessors(self.headings, accessors)
-        self._multiple_select = multiple_select
-        self._on_select = None
-        self._on_double_click = None
-        self._data = None
-        if missing_value is None:
-            print(
-                "WARNING: Using empty string for missing value in data. "
-                "Define a 'missing_value' on the table to silence this message"
+        if headings is not None:
+            self._headings = [heading.split("\n")[0] for heading in headings]
+            self._accessors = build_accessors(self._headings, accessors)
+        elif accessors is not None:
+            self._headings = None
+            self._accessors = accessors
+        else:
+            raise ValueError(
+                "Cannot create a table without either headings or accessors"
             )
+
+        self._multiple_select = multiple_select
         self._missing_value = missing_value or ""
 
+        # Prime some properties that need to exist before the table is created.
+        self.on_select = None
+        self.on_activate = None
+        self._data = None
+
         self._impl = self.factory.Table(interface=self)
         self.data = data
 
         self.on_select = on_select
-        self.on_double_click = on_double_click
+        self.on_activate = on_activate
 
     @property
-    def data(self):
-        """The data source of the widget. It accepts table data in the form of
-        ``list``, ``tuple``, or :obj:`ListSource`
+    def enabled(self) -> bool:
+        """Is the widget currently enabled? i.e., can the user interact with the widget?
+        Table widgets cannot be disabled; this property will always return True; any
+        attempt to modify it will be ignored.
+        """
+        return True
+
+    @enabled.setter
+    def enabled(self, value):
+        pass
+
+    def focus(self):
+        "No-op; Table cannot accept input focus"
+        pass
+
+    @property
+    def data(self) -> ListSource:
+        """The data to display in the table.
+
+        When setting this property:
+
+        * A :any:`Source` will be used as-is.
 
-        Returns:
-            Returns a (:obj:`ListSource`).
+        * A value of None is turned into an empty ListSource.
+
+        * Otherwise, the value must be an iterable, which is copied into a new
+          ListSource using the widget's accessors. Items are converted as shown
+          :ref:`here <listsource-item>`.
         """
         return self._data
 
     @data.setter
-    def data(self, data):
+    def data(self, data: Any):
         if data is None:
             self._data = ListSource(accessors=self._accessors, data=[])
-        elif isinstance(data, (list, tuple)):
-            self._data = ListSource(accessors=self._accessors, data=data)
-        else:
+        elif isinstance(data, Source):
             self._data = data
+        else:
+            self._data = ListSource(accessors=self._accessors, data=data)
```

Comment:
I can see what you're doing here, but I'm not 100% sure it's the right approach; details in the docs.